src/rules/switchDefaultRule.ts

```diff
@@ -36,16 +54,66 @@ export class Rule extends Lint.Rules.AbstractRule {
     public static FAILURE_STRING = "Switch statement should include a 'default' case";
 
     public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
-        return this.applyWithFunction(sourceFile, walk);
+        return this.applyWithFunction(sourceFile, walk, {
+            allowExhaustive: false, // This option requires TypeChecker
+        });
+    }
+
+    public applyWithProgram(sourceFile: ts.SourceFile, program: ts.Program): Lint.RuleFailure[] {
+        return this.applyWithFunction(
+            sourceFile,
+            walk,
+            {
+                allowExhaustive: this.ruleArguments.indexOf(OPTION_ALLOW_EXHAUSTIVE) !== -1,
+            },
+            program.getTypeChecker(),
+        );
     }
 }
 
-function walk(ctx: Lint.WalkContext<void>) {
+function walk(ctx: Lint.WalkContext<Options>, checker?: ts.TypeChecker) {
     return ts.forEachChild(ctx.sourceFile, function cb(node: ts.Node): void {
-        if (node.kind === ts.SyntaxKind.SwitchStatement &&
-            !(node as ts.SwitchStatement).caseBlock.clauses.some(isDefaultClause)) {
-            ctx.addFailureAtNode(node, Rule.FAILURE_STRING);
+        if (
+            ts.isSwitchStatement(node)
+            && !node.caseBlock.clauses.some(isDefaultClause)
+        ) {
+            const failure = !checker || !ctx.options.allowExhaustive
```

Comment:
If the option is enabled but there is no TypeChecker, IMO the rule should just print a warning and do nothing.
Otherwise users of editor plugins that cannot provide a `ts.Program` will see lint errors while editing, but they will not be present when running `tslint --project .` from CLI.